common/arena.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2025 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package common
+
+// Arena is an allocation primitive that allows individual allocations
+// out of a page of items and bulk de-allocations of last N allocations.
+// The most common way of using an Arena is to pair it up with a sync.Pool
+// of pages.
+type Arena[T any] struct {
+	used  uint32
+	pages [][]T
+
+	PageSize    uint32
+	NewPage     func() any
+	ReleasePage func(any)
+}
+
+// Alloc returns the next free item on the arena
+// Allocates a new page if needed
+func (a *Arena[T]) Alloc() *T {
+	pageIndex := a.used / a.PageSize
+	pageOffset := a.used % a.PageSize
+	if pageOffset == 0 {
```

Comment:
If `a.used` is reset to even a page earlier, then the extra page allocation becomes unnecessary.

Reply:
nice catch, thanks!